pkgs/dart_mcp/lib/src/api/tools.dart

```diff
@@ -290,7 +362,315 @@ extension type Schema.fromMap(Map<String, Object?> _value) {
   List<Schema>? get not => (_value['not'] as List?)?.cast<Schema>();
 }
 
+extension SchemaValidation on Schema {
+  /// Validates the given [data] against this schema.
+  ///
+  /// Returns a list of [ValidationError] if validation fails,
+  /// or an empty list if validation succeeds.
+  List<ValidationError> validate(Object? data) {
+    final failures = <ValidationError>[];
+    _validateSchema(data, [], failures);
+
+    // Calling .toSet().toList() isn't enough to de-duplicate, since even maps
+    // with identical members can be different maps, so we do it the hard, slow
+    // way. Duplicates arise because the same message can occur for multiple
+    // layers in the hierarchy.
+    final dedupedSet = <ValidationError>{};
+    for (final failure in failures.toSet()) {
+      var alreadyExists = false;
+      for (final dedupedFailure in dedupedSet) {
+        if (failure.path == dedupedFailure.path &&
```

Comment:
These lists will only be equal if they are identical instances, we probably need a deep comparison?